bpf/common/dns.h

```diff
@@ -158,30 +189,51 @@ static __always_inline u8 handle_dns(struct __sk_buff *skb,
         dns_req_t *req = bpf_ringbuf_reserve(&events, sizeof(dns_req_t), 0);
 
         if (req) {
-            __builtin_memcpy(&req->conn, conn, sizeof(connection_info_t));
-
-            req->flags = EVENT_DNS_REQUEST;
-            req->p_type = skb->pkt_type;
-            req->len = skb->len;
-            req->dns_q = qr;
-            req->id = bpf_ntohs(hdr.id);
-            req->ts = bpf_ktime_get_ns();
-            req->tp.ts = bpf_ktime_get_ns();
-            req->pid = conn_pid->p_info;
-
-            trace_key_t t_key = {0};
-            trace_key_from_pid_tid_with_p_key(&t_key, &conn_pid->p_key, conn_pid->id);
-
-            const u8 found = find_trace_for_client_request_with_t_key(
-                &p_conn, orig_dport, &t_key, conn_pid->id, &req->tp);
-
-            bpf_dbg_printk("handle_dns: looking up client trace info, found %d", found);
-            if (found) {
-                urand_bytes(req->tp.span_id, SPAN_ID_SIZE_BYTES);
-            } else {
-                init_new_trace(&req->tp);
-            }
-            read_skb_bytes(skb, dns_off, req->buf, sizeof(req->buf));
+            u32 len = skb->len - dns_off;
+            bpf_clamp_umax(len, 512);
+            populate_dns_record(req, &p_conn, orig_dport, len, qr, hdr.id, conn_pid);
+
+            read_skb_bytes(skb, dns_off, req->buf, len);
```

Comment:
This was a nasty bug. We were reading up to the max of req->buf, but read_skb_bytes failed to grab the last chunk of 16 bytes since we weren't specifically reading exactly as we should. This left us with garbage at the end.